Add documentation for empty window frames

aditi-pandit opened the pull request:
Adding a section on empty window frames in the window function developer guide.

PHILO-HE commented on velox/docs/develop/window.rst:
```diff
@@ -0,0 +1,198 @@
+A frame start of CURRENT ROW refers to the first peer row of the current row,
+while a frame end of CURRENT ROW refers to the last peer row of the current row.
+If no ORDER BY is specified, all rows are considered peers of the current row.
+
```
Is it necessary to mention how to handle NULL if it exists in ORDER BY column?

aditi-pandit replied:
@PHILO-HE : Only Range frames are affected by NULL values. I am planning to add another section to this doc for k range frames. I'll give more details about NULL values in that section.

aditi-pandit commented on velox/docs/develop/window.rst:
```diff
@@ -0,0 +1,198 @@
+**RANGE mode**
+
+In RANGE mode, all peer rows have the same frame number.
+Rows are peers if they have the same values for the ORDER BY fields.
```
Thats a good catch ! Thanks.

mbasmanova commented on velox/docs/develop/window.rst:
```diff
@@ -0,0 +1,202 @@
+========================
```
nit: trim these to cover just 'Window Operator'

mbasmanova commented on velox/docs/develop/window.rst:
```diff
@@ -0,0 +1,202 @@
+Window Operator
+========================
+
+Velox supports window function evaluation using Window operator. In this guide
```
> In this guide we will discuss some intricate design issues in this operator.

'issues' sounds like 'problems'. Is this intentional? 

Reading ahead it seems that this article focuses specifically on window frames. Would it make sense to state that clearly and rename the file and the doc title to reflect that?

Do you expect readers to be familiar with window functions, e.g. the content available in https://facebookincubator.github.io/velox/functions/presto/window.html ? Is this an extension of that documentation? Maybe make this clear.

BTW, where is this document referenced from?

aditi-pandit replied:
Right now, I have material on window frames since there have many details that needed a doc explanation.

I'm in 2 minds if I should write an explanation of the window function API for a developer adding a new function. In the past, we felt that it was explanatory from the code. So I named this window.rst so that it can be extended in the future if required.

Added a reference to the doc from the top level index.

mbasmanova commented on velox/docs/develop/window.rst:
```diff
@@ -0,0 +1,202 @@
+
+Window frames
+-------------
+Window functions can optionally include a FRAME clause. The FRAME clause
```
Add empty line.

mbasmanova commented on velox/docs/develop/window.rst:
```diff
@@ -0,0 +1,202 @@
+   expression FOLLOWING
+   UNBOUNDED FOLLOWING
+
+**ROWS mode**
```
This and following 2 sections seem to be repeating the information found in https://facebookincubator.github.io/velox/functions/presto/window.html. Would it make sense to enhance https://facebookincubator.github.io/velox/functions/presto/window.html instead?

aditi-pandit replied:
There is a bit of repetition but the context are entirely different. The functions doc is more declarative is saying what the frame means for the function. The develop doc is explaining what those clauses mean for the frame indices. The sentences are framed in that way.

I have now added a link from the functions doc to the window frame doc.

mbasmanova commented on velox/docs/develop/window.rst:
```diff
@@ -0,0 +1,202 @@
+
+Its possible that certain rows have empty frames. Empty frames occur when :
+
+* Both frameStart and frameEnds fall before the first partition row.
```
Should this use frame_start and frame_end for consistency?

aditi-pandit replied:
Yeah, made the usage consistent across the writing.

mbasmanova commented on velox/docs/develop/window.rst:
```diff
@@ -0,0 +1,202 @@
+(or precede) each other. Adhoc frames that use column values for bounds can have
+empty, partial or valid frames at any points in the partition rows.
+
+**Handling empty frames in window functions**
```
Wouldn't it make sense to document the semantics of empty frames in the user-visible docs: https://facebookincubator.github.io/velox/functions/presto/window.html ?

aditi-pandit replied:
That doc seems more applicable for the functions itself. It is heavily borrowed from https://prestodb.io/docs/0.283/functions/window.html. It might be better to link the current doc from that one. I have added a link now.

mbasmanova commented on velox/docs/develop/window.rst:
```diff
@@ -0,0 +1,202 @@
+The most naive approach to handle empty frames is to check in the window
+function logic if the frame indices are an empty frame (based on the
+conditions previously described) and return the null output.
+This is very inefficient.
```
Wondering why is this inefficient.

aditi-pandit replied:
If there are multiple window functions in the same operator this logic would be evaluated multiple times. Also the code is repeated in every function. Have clarified in the doc.

mbasmanova reviewed (commented):
@aditi-pandit Aditi, thank you for working on improving documentation for window functions. It feels like this new doc contains a mix of user-facing documentation that clarifies the semantics of empty and partial frames and some implementation specific details of the Window operator. I wonder if it makes sense to move user-facing details to https://facebookincubator.github.io/velox/functions/presto/window.html and implementation details to the comments in the code.

aditi-pandit commented:
> @aditi-pandit Aditi, thank you for working on improving documentation for window functions. It feels like this new doc contains a mix of user-facing documentation that clarifies the semantics of empty and partial frames and some implementation specific details of the Window operator. I wonder if it makes sense to move user-facing details to https://facebookincubator.github.io/velox/functions/presto/window.html and implementation details to the comments in the code.

@mbasmanova :
https://facebookincubator.github.io/velox/functions/presto/window.html and also https://facebookincubator.github.io/velox/functions/spark/window.html in turn are mainly for the functions. I feel adding a link from the functions doc to the window frames doc is better. wdyt ?

I wrote this doc somewhat in line with join and anti-join docs since they too provide many examples with some sense of the code. Since empty frames influence the WindowFunction API, I felt it should be mentioned. That part is more developer facing than user-facing. But since Velox users are mostly developers the boundaries are vague.

aditi-pandit commented:
@mbasmanova : Thanks for the review. Have addressed your comments.

aditi-pandit commented:
@mbasmanova : We now have a window function developer guide at https://facebookincubator.github.io/velox/develop/window.html.

This PR enhances the docs for the usage of valid_frames argument in it.

Please can you review.

kagamiori commented on velox/docs/develop/window.rst:
```diff
@@ -205,3 +205,107 @@ peer row with a NULL and the frame_end index is the last peer row with a NULL va
+
+A partial frame occurs when:
+
+- frame_start < frame_end (so the empty frame condition is not violated)
```
> so the empty frame condition is not violated

nit: this reads like the condition for empty frame is satisfied so that the current frame is empty. Maybe change it to "so it's not an empty frame"?

aditi-pandit replied:
Good idea. Done.

kagamiori commented on velox/docs/develop/window.rst:
```diff
@@ -205,3 +205,107 @@ peer row with a NULL and the frame_end index is the last peer row with a NULL va
+  frame_end within the partition, or frame_start is in the partition but
+  frame_end beyond it.
+
+For such frames the violating boundary is clamped to the first or last row
+of the partition.
```
Could we clarify this a bit by discussing the two cases separately? E.g., frame_start before the partition bound should be clamped to the first row of the partition. frame_end after the partition bound should be clamped to the last row of the partition.

aditi-pandit replied:
Done.

kagamiori commented on velox/docs/develop/window.rst:
```diff
@@ -205,3 +205,107 @@ peer row with a NULL and the frame_end index is the last peer row with a NULL va
+So the empty frames, partial frames and valid frames cluster together and follow
+(or precede) each other.
+
+Adhoc frames that use column values for bounds can have
```
Could we give an example of adhoc frames?

aditi-pandit replied:
Done.

kagamiori reviewed (commented):
Hi @aditi-pandit, thank you for adding the documentation! I left a few comments.

aditi-pandit commented:
> Hi @aditi-pandit, thank you for adding the documentation! I left a few comments.
> 
> For the figure, could you add the index of the current row for each of the example frame so make it more clear that they're concrete frames during window sliding?

Thanks @kagamiori for your review.

Have addressed your review comments. PTAL.

kagamiori commented on velox/docs/develop/window.rst:
```diff
@@ -205,3 +205,109 @@ peer row with a NULL and the frame_end index is the last peer row with a NULL va
+partition, but 2 preceding frame_end is within the partition. So for these
+3 rows frame_start is clamped to the first partition row.
+
+Similarly for frame *ROWS BETWEEN 2 FOLLOWING AND 5 FOLLOWING*, the 3-5th rows
+have frame_start within the partition, but frame_end beyond. So they are partial
```
It should be the last 3-5th rows here?

aditi-pandit replied:
Yes. Nice catch. Thanks.

kagamiori commented on velox/docs/develop/window.rst:
```diff
@@ -205,3 +205,109 @@ peer row with a NULL and the frame_end index is the last peer row with a NULL va
+       vector_size_t resultOffset,
+       const VectorPtr& result) = 0;
+
+The Window operator also clamps *partial* window frame indices to
+the first or final partition row. So the Window function doesn't need
+any special logic for partial frames.
```
nit: The Window operator also clamps *partial* window frame indices to
the first or final partition row before passing them to the Window function. So the Window function doesn't...

aditi-pandit replied:
+1

kagamiori reviewed (commented):
Hi @aditi-pandit, thank you for iterating on this PR. It looks good to me except a few nits.

aditi-pandit commented:
> Hi @aditi-pandit, thank you for iterating on this PR. It looks good to me except a few nits.

Thanks @kagamiori. Have addressed the review comments.

kagamiori commented on velox/docs/develop/window.rst:
```diff
@@ -205,3 +205,109 @@ peer row with a NULL and the frame_end index is the last peer row with a NULL va
 Rows with NULL values do not participate in the frames of the other rows.
 
+
+Empty frames
```
nit: Maybe "Empty and partial frames" since we talk about partial frames in this section too?

aditi-pandit replied:
@kagamiori : So the thing is that the function author needs to understand the concept of empty frames and the corresponding validFrames function parameter in the API. 

Partial frames need to be called out. But since the framework handles them the function author as such doesn't need to understand or handle them. 

So I feel its better to call this section "Empty frames" and have a sub-section for partial frames.

kagamiori replied:
@aditi-pandit, then could you clarify at the beginning of this section that empty frames need to be handled explicitly by window function authors while partial frames do not? Also, could we put the subsection "Handling empty frames in window functions" together with the introduction of empty frames, and put the subsection "Partial frames" after that?

aditi-pandit replied:
@kagamiori : I did some rewording where I introduced the type of frames in the beginning stating that empty frames need special handling. Then the section goes on to describe the frame types and the final section has instructions on how to handle empty frames.

The Empty frames heading is inline with other headings in this doc Window frames, k Range frames and empty frames.

kagamiori replied:
@aditi-pandit, is it possible to move the "Partial frames" subsection after the "Handling empty frames in window functions" subsection? I think it would be clearer to talk about empty frames and handling them in window functions, then partial frames and that they don't need special handling in window functions. What do you think?

aditi-pandit replied:
@kagamiori : There is the diagram which visually shows all 3 so I felt its better to have all the definitions and then the handling section at the end. Else the diagram shows partial frames without any definition for it.

The last line of the Handling emtpy frames section also states that there is no special handling for partial frames again. So it is self-contained as well.

kagamiori replied:
OK, this makes sense.

kagamiori reviewed (approved):
LGTM. Thank you for adding the documentation!
